src/flamenco/runtime/fd_blockstore.c

```diff
@@ -190,7 +193,55 @@ static inline void check_read_write_err( int err ) {
   }
 }
 
-/* build the archival file index */
+bool 
+fd_blockstore_invalid_archive( fd_blockstore_t * blockstore, fd_blockstore_archive_metadata_t * metadata ) {
+  return ( metadata->lrw_off < FD_BLOCKSTORE_ARCHIVE_START ) 
+         || ( metadata->mrw_end < FD_BLOCKSTORE_ARCHIVE_START )
+         || ( metadata->fd_size_max != blockstore->fd_size_max ); // should be initialized same as archive file
+}
+
+#define check_read_err_safe( cond, msg )          \
+  if( FD_UNLIKELY( cond ) ) {                     \
+    FD_LOG_WARNING(( "[%s] %s", __func__, msg )); \
+    return FD_BLOCKSTORE_ERR_SLOT_MISSING;        \
+  }
+
+/* Where read_off is where to start reading from */
+/* Guarantees that read_off is at the end of what we just finished on return. */
+static int read_with_wraparound( fd_blockstore_t * blockstore, 
+                                 int fd, 
+                                 uchar * dst, 
+                                 ulong dst_sz, 
+                                 ulong * rsz, 
+                                 ulong * read_off ) {
+  check_read_err_safe( lseek( fd, (long)*read_off, SEEK_SET ) == -1, 
+                       "failed to seek to read offset" );
+  
+  ulong remaining_sz = blockstore->fd_size_max - *read_off;
+  if ( remaining_sz < dst_sz ) {
+    int err = fd_io_read( fd, dst, remaining_sz, remaining_sz, rsz );
+    check_read_err_safe( err, "failed to read file near end" );
```

Comment:
if remaining_sz != *rsz, it is also an error

Reply:
I believe [fd_io_read already returns](https://github.com/firedancer-io/firedancer/blob/4aa0d7b17c40e433875aa7fb65a7efc71507ff2f/src/util/io/fd_io.h#L251) an errcode < 0 if rsz < remaining_sz?

Reply:
if the file is shorter than expected or we run out of disk space, read/write will return a smaller value

Reply:
read/write returns a smaller value in the rsz, but the err code returned is negative, which is checked in `check_read_err_safe` -- is this what you are referring to?